exercises/chapter11/test/no-peeking/Solutions.purs

```diff
@@ -0,0 +1,116 @@
+module Test.NoPeeking.Solutions where
+
+import Prelude
+
+import Control.Bind (bindFlipped)
+import Control.Monad.Except (ExceptT, throwError)
+import Control.Monad.Reader (Reader, ReaderT, ask, lift, local, runReader, runReaderT)
+import Control.Monad.State (State, StateT, get, put, modify, execState)
+import Control.Monad.Writer (Writer, WriterT, tell, runWriter, execWriterT)
+import Data.Array (filter, any)
+import Data.Identity (Identity)
+import Data.Maybe (Maybe(..))
+import Data.Monoid (power)
+import Data.Monoid.Additive (Additive(..))
+import Data.Newtype (unwrap)
+import Data.String (joinWith)
+import Data.String.CodeUnits (stripPrefix, toCharArray)
+import Data.String.Pattern (Pattern(..))
+import Data.Traversable (sequence, traverse_)
+import Data.Tuple (Tuple)
+
+--
+
+testParens :: String -> Boolean
+testParens str = 
+  let
+    sanitized = str 
+      # toCharArray 
+      # filter (any (==) ['(', ')'])
+      # map (\c -> if c == '(' then 1 else -1)
+
+    process :: Array Int -> State (Maybe Int) Unit
+    process = traverse_ 
+      \n -> modify $ bindFlipped 
+        \st -> if st + n >= 0 
+          then Just (st + n) 
+          else Nothing
+  in
+  case execState (process sanitized) (Just 0) of
+    Just 0 -> true
+    _ -> false
```

Comment:
For reference, here's what I came up with a while back. Feel free to use whatever parts of each version you find clearest:
```purs
testParens :: String -> Boolean    
testParens str =       
  let        
    ca = toCharArray str    
    
    openTally :: Char -> Int -> Int    
    -- Open parens only considered if not already in deficit    
    -- No recovery from too-many closed parens    
    openTally '(' tally | tally >= 0 = tally + 1    
    openTally ')' tally = tally - 1    
    -- Non-parens has no effect    
    openTally _ tally = tally    
    
    sumParens :: Array Char -> State Int Unit    
    -- Either modify or modify_ are fine    
    --sumParens = traverse_ \c -> modify $ openTally c    
    sumParens = traverse_ \c -> modify_ $ openTally c    
    
    finalTally :: Int    
    finalTally = execState (sumParens ca) 0    
  in    
    finalTally == 0
```

Reply:
I used your solution as it's more straightforward. Thanks